popde/adaptive_kde.py

```diff
@@ -224,6 +224,143 @@ def optimize_parameters(self, method='loo_cv', fom_plot_name=None):
+    score, based on log-likelihood as the figure of merit.
+
+    Attributes:
+        alpha (float): Initial alpha parameter for rescaling.
+        bandwidth (float): Fixed bandwidth, default set to 1.0.
+        n_splits (int): Number of splits for k-fold cross-validation.
+
+    Methods:
+        loo_cv_score(rescale_factors_alpha): 
+            Computes Leave-One-Out cross-validation score 
+            for given rescale factors.
+        kfold_cv_score(rescale_factors_alpha, seed=42): 
+            Computes k-fold cross-validation score 
+            for given rescale factors.
+        optimize_rescale_parameters(initial_rescale_factor, initial_alpha=0.0, method='loo_cv', fom_plot_name=None, bounds=None):
+            Optimizes rescaling factors and alpha parameter.
+    """
+    def __init__(self, data, rescale_factors_and_alpha_array, weights=None, input_transf=None,
+                 stdize=False, rescale=None, backend='KDEpy', bandwidth=1.0, alpha=0.0,
+                 dim_names=None, do_fit=False, n_splits=2):
+        """
+        Initialize the KDERescaleOptimization instance 
+        with data and optional configuration parameters.
+
+        Args:
+            data (array-like): Data to fit the KDE model.
+            rescale_factors_and_alpha_array (array-like): Initial rescale factors
+               and alpha parameter array.
+            weights (array-like, optional): Weights for each data point.
+            input_transf (function, optional): Input transformation function.
+            stdize (bool, optional): Whether to standardize data.
+            rescale (array-like, optional): Initial rescale factors 
+               for each dimension.
+            backend (str, optional): Backend to use, default is 'KDEpy'.
+            bandwidth (float, optional): Fixed bandwidth value, default is 1.0.
+            alpha (float, optional): Initial alpha value, default is 0.0.
+            dim_names (list, optional): Dimension names.
+            do_fit (bool, optional): Whether to fit the KDE upon initialization.
+            n_splits (int, optional): Number of splits for k-fold 
+               cross-validation, default is 2.
+        """
+
+        self.n_splits = n_splits
+        #not sure I need this
+        self.alpha = alpha
+        self.bandwidth = bandwidth
+
+        super().__init__(data, weights, input_transf, stdize, rescale, backend, 
+                bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, rescale_factors_alpha):
+        """
+        Compute the Leave-One-Out Cross-Validation (LOO-CV) score
+          based on log-likelihood.
+
+        Args:
+            rescale_factors_alpha (array-like): Array of rescale factors per
+              dimension and alpha value.
+
+        Returns:
+            float: Negative sum of log-likelihood scores for LOO-CV.
+        """
+        alpha = rescale_factors_alpha[-1]
+        rescale_factor = rescale_factors_alpha[0:len(rescale_factors_alpha)-1]
```

Comment:
can you please look up how to index on the last value of an array or the last but 1 value (etc), numpy allows you to do that with a simple syntax

Reply:
(that is if you really need to keep this rescale-plus-alpha array which seems a bit awkward)